src/precompiles/mod.rs

```diff
@@ -168,20 +156,99 @@ pub fn berlin_precompiles(
         None => return Some(PrecompileResult::Err(ExitError::OutOfGas)),
     };
 
-    match address.to_low_u64_be() {
-        1 => Some(ECRecover::run(input, target_gas, context)),
-        2 => Some(SHA256::run(input, target_gas, context)),
-        3 => Some(RIPEMD160::run(input, target_gas, context)),
-        4 => Some(Identity::run(input, target_gas, context)),
-        5 => Some(ModExp::<Berlin>::run(input, target_gas, context)), // TODO gas changes
-        6 => Some(BN128Add::<Istanbul>::run(input, target_gas, context)),
-        7 => Some(BN128Mul::<Istanbul>::run(input, target_gas, context)),
-        8 => Some(BN128Pair::<Istanbul>::run(input, target_gas, context)),
-        9 => Some(Blake2F::run(input, target_gas, context)),
+    match address.0 {
+        ECRecover::ADDRESS => Some(ECRecover::run(input, target_gas, context)),
+        SHA256::ADDRESS => Some(SHA256::run(input, target_gas, context)),
+        RIPEMD160::ADDRESS => Some(RIPEMD160::run(input, target_gas, context)),
+        Identity::ADDRESS => Some(Identity::run(input, target_gas, context)),
+        modexp::ADDRESS => Some(ModExp::<Berlin>::run(input, target_gas, context)), // TODO gas changes
+        bn128::addresses::ADD => Some(BN128Add::<Istanbul>::run(input, target_gas, context)),
+        bn128::addresses::MUL => Some(BN128Mul::<Istanbul>::run(input, target_gas, context)),
+        bn128::addresses::PAIR => Some(BN128Pair::<Istanbul>::run(input, target_gas, context)),
+        Blake2F::ADDRESS => Some(Blake2F::run(input, target_gas, context)),
         #[cfg(feature = "contract")]
-        EXIT_TO_NEAR_ID => Some(ExitToNear::run(input, target_gas, context)),
+        ExitToNear::ADDRESS => Some(ExitToNear::run(input, target_gas, context)),
         #[cfg(feature = "contract")]
-        EXIT_TO_ETHEREUM_ID => Some(ExitToEthereum::run(input, target_gas, context)),
+        ExitToEthereum::ADDRESS => Some(ExitToEthereum::run(input, target_gas, context)),
         _ => None,
     }
 }
+
+/// const fn for making an address by concatenating the bytes from two given numbers,
+/// Note that 32 + 128 = 160 = 20 bytes (the length of an address). This function is used
+/// as a convenience for specifying the addresses of the various precompiles.
+const fn make_address(x: u32, y: u128) -> [u8; 20] {
+    let x_bytes = x.to_be_bytes();
+    let y_bytes = y.to_be_bytes();
+    [
+        x_bytes[0],
+        x_bytes[1],
+        x_bytes[2],
+        x_bytes[3],
+        y_bytes[0],
+        y_bytes[1],
+        y_bytes[2],
+        y_bytes[3],
+        y_bytes[4],
+        y_bytes[5],
+        y_bytes[6],
+        y_bytes[7],
+        y_bytes[8],
+        y_bytes[9],
+        y_bytes[10],
+        y_bytes[11],
+        y_bytes[12],
+        y_bytes[13],
+        y_bytes[14],
+        y_bytes[15],
+    ]
```

Comment:
After I wrote this, I learned that `loop` and `if` are allowed in `const fn`. This means an alternate implementation would be:

```rust
    let mut i = 0;
    let mut bytes = [0u8; 20];
    loop {
        if i < 4 {
            bytes[i] = x_bytes[i];
        } else if i < 20 {
            bytes[i] = y_bytes[i - 4];
        } else {
            break;
        }
        i += 1;
    }

    bytes
```

Note: `for` and `&mut` (and therefore `copy_from_slice`) are currently not allowed in `const fn`.

The `loop` implementation is less boilerplate-y than the current implementation, but arguably harder to read (more cognitive overhead). I think it takes a little longer to compile too (note this function is only used at compile-time, so its efficiency only impacts the compilation of the project), which makes sense because the loop has a lot more going on than just copying bytes from one place to another (the arithmetic of tracking the index, the conditional figuring out if it should take from `x_bytes` or `y_bytes`, etc.)

So ultimately I think the current implementation is better than the `loop` one, but I thought I would write this up for academic purposes.

Reply:
Yeah, I also noticed this part which looked a little bit strange for me, but I knew that if you did it, there was a reason for that. So immediately tried different approaches locally and ended up with the same conclusions, that I can't make it look prettier, rather harder to read :)

So I'm also for keeping it as the same. As approaches that work for `const fn` for this case are quite ugly.